lib/sshagent/client_test.go

```diff
@@ -0,0 +1,125 @@
+// Teleport
+// Copyright (C) 2025 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package sshagent_test
+
+import (
+	"context"
+	"io"
+	"net"
+	"path/filepath"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/crypto/ssh/agent"
+
+	"github.com/gravitational/teleport/lib/sshagent"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+func TestSSHAgentClient(t *testing.T) {
+	keyring, ok := agent.NewKeyring().(agent.ExtendedAgent)
+	require.True(t, ok)
+
+	agentDir := t.TempDir()
+	agentPath := filepath.Join(agentDir, "agent.sock")
+	startAgentServer := func() (stop func()) {
+		l, err := net.Listen("unix", agentPath)
+		require.NoError(t, err)
+
+		// create a context to close existing connections on server shutdown.
+		serveCtx, serveCancel := context.WithCancel(t.Context())
+
+		// Track open connections.
+		var connWg sync.WaitGroup
+
+		go func() {
```

Comment:
You can only call Add if you're "guarded" by an existing add or if there's no way for anyone to have called Wait yet - with the code as is you can get this sequence of events:

- `l.Accept` returns a connection in this goroutine
- `l.Close` and `serveCancel` are called in `stopServer`
-  `connWg.Add` in this goroutine is called concurrently with `connWg.Wait` in `stopServer`, which blows up

```suggestion
		var connWg sync.WaitGroup

		connWg.Add(1)
		go func() {
			defer connWg.Done()
```